File: packages/iroh_quic/lib/src/endpoint.dart
/// Endpoint lifecycle + connections. [Endpoint.bind] is the first async surface;
/// [Endpoint.connect] establishes a [Connection]. This library is composed from `connection.dart`
/// and `stream.dart` (as parts) so the wrapper types can share private constructors. Call
/// [Iroh.init] once before binding.
library;

import 'dart:async';
import 'dart:typed_data';

import 'package:flutter_rust_bridge/flutter_rust_bridge.dart' show AnyhowException;

import 'errors.dart';
import 'key.dart';
import 'rust/api/address_lookup.dart' as ffi_addr;
import 'rust/api/connection.dart' as ffi_net;
import 'rust/api/endpoint.dart' as ffi_ep;
import 'rust/api/key.dart' as ffi_key;
import 'rust/api/router.dart' as ffi_router;
import 'rust/api/watch.dart' as ffi_watch;
import 'watch.dart';

part 'connection.dart';
part 'router.dart';
part 'stream.dart';

EndpointAddr _addrFromParts(ffi_key.EndpointAddrParts parts) => EndpointAddr(
  PublicKey.fromBytes(parts.id),
  relayUrls: parts.relayUrls.map(RelayUrl.parse).toList(),
  ipAddrs: parts.ipAddrs,
);

int _streamTokenCounter = 0;

/// Wraps an FRB reactive stream (which blocks a Rust task that cannot otherwise observe Dart
/// cancellation) so that cancelling the returned subscription aborts the Rust task via
/// `cancelStream(token)` - making cancel prompt and leak-free.
Stream<R> _cancellableStream<T, R>(
  Stream<T> Function(BigInt token) create,
  R Function(T value) convert,
) {
  final token = BigInt.from(++_streamTokenCounter);
  late StreamController<R> controller;
  StreamSubscription<T>? sub;
  controller = StreamController<R>(
    onListen: () {
      sub = create(token).listen(
        (value) => controller.add(convert(value)),
        onError: controller.addError,
        onDone: controller.close,
      );
    },
    onCancel: () async {
      ffi_watch.cancelStream(token: token);
      await sub?.cancel();
    },
  );
  return controller.stream;
}

/// A bound iroh endpoint - the local participant in the P2P network. Holds native runtime
/// resources (sockets, tokio tasks, relay connections); call [close] to release them.
final class Endpoint {
  Endpoint._(this._inner);

  final ffi_ep.Endpoint _inner;

  /// Binds a new endpoint and returns once it is ready to connect/accept.
  ///
  /// [secretKey] fixes the identity (a fresh random key is generated if omitted). [alpns] are the
  /// application protocols accepted for inbound connections. [relayMode] selects relay servers
  /// (defaults to n0's production relays). Throws [IrohBindException] on failure.
  static Future<Endpoint> bind({
    SecretKey? secretKey,
    List<List<int>> alpns = const <List<int>>[],
    RelayMode relayMode = RelayMode.n0Default,
  }) async {
    final (kind, customUrls) = _relayModeArgs(relayMode);
    try {
      final inner = await ffi_ep.endpointBind(
        secretKey: secretKey == null ? null : Uint8List.fromList(secretKey.toBytes()),
        alpns: alpns.map(Uint8List.fromList).toList(),
        relayModeKind: kind,
        customRelayUrls: customUrls,
      );
      return Endpoint._(inner);
    } on AnyhowException catch (e) {
      throw IrohBindException(e.message, cause: e);
    }
  }

  /// Binds an endpoint that resolves peer addresses via a **Dart-implemented** [resolve] callback
  /// instead of iroh's built-in DNS/pkarr lookup. When you `connect` to a peer
  /// known only by its [EndpointId], [resolve] is invoked to discover its [EndpointAddr]; return
  /// `null` for unknown ids. Throws [IrohBindException] on failure.
  static Future<Endpoint> bindWithAddressLookup({
    required FutureOr<EndpointAddr?> Function(EndpointId endpointId) resolve,
    SecretKey? secretKey,
    List<List<int>> alpns = const <List<int>>[],
    RelayMode relayMode = RelayMode.n0Default,
  }) async {
    final (kind, customUrls) = _relayModeArgs(relayMode);
    try {
      final inner = await ffi_addr.endpointBindWithAddressLookup(
        secretKey: secretKey == null ? null : Uint8List.fromList(secretKey.toBytes()),
        alpns: alpns.map(Uint8List.fromList).toList(),
        relayModeKind: kind,
        customRelayUrls: customUrls,
        resolve: (Uint8List idBytes) async {
          final addr = await resolve(PublicKey.fromBytes(idBytes));
          if (addr == null) return null;
          return ffi_key.EndpointAddrParts(
            id: addr.id.asBytes(),
            relayUrls: addr.relayUrls.map((u) => u.value).toList(),
            ipAddrs: addr.ipAddrs,
          );
        },
      );
      return Endpoint._(inner);
    } on AnyhowException catch (e) {
      throw IrohBindException(e.message, cause: e);
    }
  }

  /// This endpoint's [EndpointId] (public key).
  EndpointId get id => PublicKey.fromBytes(_inner.id());

  /// A snapshot of this endpoint's current address (id + known relay URLs + direct IP addrs).
  /// The relay/direct lists may be empty immediately after [bind] until discovery populates them.
  EndpointAddr get addr => _addrFromParts(_inner.addr());

  /// A live [Stream] of this endpoint's [EndpointAddr] as relay/direct discovery updates it.
  /// Cancelling the subscription drops the iroh watcher.
  Stream<EndpointAddr> watchAddr() =>
      _cancellableStream((token) => _inner.watchAddr(token: token), _addrFromParts);

  /// A live [Stream] of the home-relay connection status (one [RelayStatus] per configured relay).
  Stream<List<RelayStatus>> homeRelayStatus() => _cancellableStream(
    (token) => _inner.homeRelayStatus(token: token),
    (list) => list
        .map((s) => RelayStatus(url: s.url, connected: s.connected, lastError: s.lastError))
        .toList(),
  );

  /// The local socket addresses (`host:port`) this endpoint is bound to.
  List<String> get boundSockets => _inner.boundSockets();

  /// Whether the endpoint has been closed.
  bool get isClosed => _inner.isClosed();

  /// Replaces the accepted ALPN protocols (affects new inbound connections only).
  void setAlpns(List<List<int>> alpns) =>
      _inner.setAlpns(alpns: alpns.map(Uint8List.fromList).toList());

  /// Connects to [addr] over the [alpn] protocol, returning an established [Connection].
  /// Throws [IrohConnectException] if the connection cannot be established.
  Future<Connection> connect(EndpointAddr addr, List<int> alpn) async {
    final parts = ffi_key.EndpointAddrParts(
      id: addr.id.asBytes(),
      relayUrls: addr.relayUrls.map((u) => u.value).toList(),
      ipAddrs: addr.ipAddrs,
    );
    try {
      final inner = await _inner.connect(addr: parts, alpn: alpn);
      return Connection._(inner);
    } on AnyhowException catch (e) {
      throw IrohConnectException(e.message, cause: e);
    }
  }

  /// Accepts the next inbound [Connection], or `null` once the endpoint is closed. Collapses
  /// iroh's lazy two-step accept. Throws [IrohAcceptException] on failure.
  Future<Connection?> accept() async {
    try {
      final inner = await _inner.accept();
      return inner == null ? null : Connection._(inner);
    } on AnyhowException catch (e) {
      throw IrohAcceptException(e.message, cause: e);
    }
  }

  /// Accepts the next inbound connection as an unaccepted [Incoming], letting you inspect the
  /// remote address and choose `accept`/`refuse`/`retry`/`ignore`. Returns `null` once
  /// the endpoint is closed. Throws [IrohAcceptException] on failure.
  Future<Incoming?> acceptIncoming() async {
    try {
      final inner = await _inner.acceptIncoming();
      return inner == null ? null : Incoming._(inner);
    } on AnyhowException catch (e) {
      throw IrohAcceptException(e.message, cause: e);
    }
  }

  /// Begins building a multi-protocol [Router] over this endpoint. Register one Dart
  /// handler per ALPN, then `spawn()`. Use this for in-process multiplexing of several protocols;
  /// for the simple single-protocol case prefer the [accept] loop.
  Future<RouterBuilder> router() async =>
      RouterBuilder._(await ffi_router.routerBuilder(endpoint: _inner));

  /// Closes the endpoint gracefully, tearing down connections and background tasks.
  Future<void> close() => _inner.close();
}

(int, List<String>) _relayModeArgs(RelayMode mode) => switch (mode) {
  RelayModeDefault() => (0, const <String>[]),
  RelayModeDisabled() => (1, const <String>[]),
  RelayModeStaging() => (2, const <String>[]),
  RelayModeCustom(:final map) => (3, map.urls.map((u) => u.value).toList()),
};

File: packages/iroh_flutter/rust/src/api/endpoint.rs
//! Endpoint lifecycle - the first async surface.
//!
//! `Endpoint` is an **opaque** handle (it owns iroh's runtime resources - sockets, tokio tasks,
//! relay connections). `Endpoint::builder(presets::N0)` + the rustls `CryptoProvider` (installed
//! by the `tls-ring` feature) are hidden entirely behind [`endpoint_bind`]. Every async op runs on
//! the shared tokio runtime (`crate::runtime::rt`).

use std::str::FromStr;

use anyhow::{anyhow, Result};
use flutter_rust_bridge::frb;
use futures::StreamExt;
use iroh::endpoint::presets;
use iroh::{Endpoint as IrohEndpoint, RelayMode, RelayUrl, SecretKey, Watcher};

use crate::api::connection::{Connection, Incoming};
use crate::api::key::{build_endpoint_addr, parts_from_endpoint_addr, to_arr32, EndpointAddrParts};
use crate::api::watch::RelayStatusInfo;
use crate::frb_generated::StreamSink;
use crate::runtime::rt;

/// A bound iroh endpoint. Opaque handle; methods dispatch into iroh on the shared tokio runtime.
#[frb(opaque)]
pub struct Endpoint {
    pub(crate) inner: IrohEndpoint,
}

/// Maps the Dart-side relay-mode discriminant (+ custom URLs) onto iroh's `RelayMode`.
/// 0 = Default, 1 = Disabled, 2 = Staging, 3 = Custom.
pub(crate) fn map_relay_mode(kind: i32, custom: Vec<String>) -> Result<RelayMode> {
    Ok(match kind {
        0 => RelayMode::Default,
        1 => RelayMode::Disabled,
        2 => RelayMode::Staging,
        3 => {
            let urls = custom
                .iter()
                .map(|u| RelayUrl::from_str(u).map_err(|e| anyhow!("{e}")))
                .collect::<Result<Vec<RelayUrl>>>()?;
            RelayMode::custom(urls)
        }
        _ => return Err(anyhow!("invalid relay mode discriminant: {kind}")),
    })
}

/// Binds a new endpoint. Hides `presets::N0` (DNS/pkarr address lookup + relays + crypto provider);
/// `relay_mode_kind`/`custom_relay_urls` override the relay configuration, `alpns` declares the
/// protocols accepted for inbound connections, and `secret_key` (if given) fixes the identity.
#[frb]
pub async fn endpoint_bind(
    secret_key: Option<Vec<u8>>,
    alpns: Vec<Vec<u8>>,
    relay_mode_kind: i32,
    custom_relay_urls: Vec<String>,
) -> Result<Endpoint> {
    let relay_mode = map_relay_mode(relay_mode_kind, custom_relay_urls)?;
    let sk = match secret_key {
        Some(bytes) => Some(SecretKey::from_bytes(&to_arr32(&bytes)?)),
        None => None,
    };
    let inner = rt()
        .spawn(async move {
            let mut builder = IrohEndpoint::builder(presets::N0)
                .relay_mode(relay_mode)
                .alpns(alpns);
            if let Some(sk) = sk {
                builder = builder.secret_key(sk);
            }
            let ep = builder.bind().await?;
            Ok::<IrohEndpoint, anyhow::Error>(ep)
        })
        .await
        .map_err(|e| anyhow!("endpoint bind task panicked: {e}"))??;
    Ok(Endpoint { inner })
}

impl Endpoint {
    /// The endpoint's [`EndpointId`](iroh::EndpointId) (public key) as 32 raw bytes.
    #[frb(sync)]
    pub fn id(&self) -> Vec<u8> {
        self.inner.id().as_bytes().to_vec()
    }

    /// A snapshot of this endpoint's current address (id + known relay URLs + direct IP addrs).
    #[frb(sync)]
    pub fn addr(&self) -> EndpointAddrParts {
        parts_from_endpoint_addr(&self.inner.addr())
    }

    /// The local socket addresses this endpoint is bound to (`host:port`).
    #[frb(sync)]
    pub fn bound_sockets(&self) -> Vec<String> {
        self.inner
            .bound_sockets()
            .iter()
            .map(|s| s.to_string())
            .collect()
    }

    /// Replaces the set of accepted ALPN protocols (affects new inbound connections only).
    #[frb(sync)]
    pub fn set_alpns(&self, alpns: Vec<Vec<u8>>) {
        self.inner.set_alpns(alpns);
    }

    /// Whether the endpoint has been closed.
    #[frb(sync)]
    pub fn is_closed(&self) -> bool {
        self.inner.is_closed()
    }

    /// Connects to a remote endpoint over the given ALPN protocol, returning an established
    /// [`Connection`]. Mirrors `iroh::Endpoint::connect`.
    pub async fn connect(&self, addr: EndpointAddrParts, alpn: Vec<u8>) -> Result<Connection> {
        let target = build_endpoint_addr(&addr)?;
        let ep = self.inner.clone();
        let conn = rt()
            .spawn(async move {
                let conn = ep.connect(target, &alpn).await?;
                Ok::<_, anyhow::Error>(conn)
            })
            .await
            .map_err(|e| anyhow!("connect task panicked: {e}"))??;
        Ok(Connection::wrap(conn))
    }

    /// Accepts the next inbound connection, collapsing the lazy `Accept -> Incoming -> Accepting ->
    /// Connection` chain. Resolves to `None` once the endpoint is closed.
    pub async fn accept(&self) -> Result<Option<Connection>> {
        let ep = self.inner.clone();
        let maybe = rt()
            .spawn(async move {
                let Some(incoming) = ep.accept().await else {
                    return Ok::<Option<Connection>, anyhow::Error>(None);
                };
                let connecting = incoming.accept()?;
                let conn = connecting.await?;
                Ok(Some(Connection::wrap(conn)))
            })
            .await
            .map_err(|e| anyhow!("accept task panicked: {e}"))??;
        Ok(maybe)
    }

    /// Accepts the next inbound connection as an unaccepted [`Incoming`], letting the caller
    /// inspect the remote address and choose `accept`/`refuse`/`retry`/`ignore`. Resolves to
    /// `None` once the endpoint is closed.
    pub async fn accept_incoming(&self) -> Result<Option<Incoming>> {
        let ep = self.inner.clone();
        let maybe = rt()
            .spawn(async move {
                Ok::<Option<iroh::endpoint::Incoming>, anyhow::Error>(ep.accept().await)
            })
            .await
            .map_err(|e| anyhow!("accept_incoming task panicked: {e}"))??;
        Ok(maybe.map(Incoming::wrap))
    }

    /// Streams snapshots of this endpoint's [`EndpointAddr`](iroh::EndpointAddr) as relay/direct
    /// discovery updates it. `token` lets the Dart subscription cancel the underlying task via
    /// `cancel_stream`, which drops the iroh watcher - no leak.
    pub fn watch_addr(&self, token: u64, sink: StreamSink<EndpointAddrParts>) {
        let watcher = self.inner.watch_addr();
        let handle = rt().spawn(async move {
            let mut stream = watcher.stream();
            while let Some(addr) = stream.next().await {
                if sink.add(parts_from_endpoint_addr(&addr)).is_err() {
                    break;
                }
            }
            crate::runtime::unregister_stream(token);
        });
        crate::runtime::register_stream(token, handle.abort_handle());
    }

    /// Streams the home-relay connection status (one entry per configured relay) as it changes.
    pub fn home_relay_status(&self, token: u64, sink: StreamSink<Vec<RelayStatusInfo>>) {
        let watcher = self.inner.home_relay_status();
        let handle = rt().spawn(async move {
            let mut stream = watcher.stream();
            while let Some(statuses) = stream.next().await {
                let mapped: Vec<RelayStatusInfo> = statuses
                    .iter()
                    .map(|s| RelayStatusInfo {
                        url: s.url().to_string(),
                        connected: s.is_connected(),
                        last_error: s.last_error().map(|e| e.to_string()),
                    })
                    .collect();
                if sink.add(mapped).is_err() {
                    break;
                }
            }
            crate::runtime::unregister_stream(token);
        });
        crate::runtime::register_stream(token, handle.abort_handle());
    }

    /// Closes the endpoint gracefully, tearing down connections and background tasks.
    pub async fn close(&self) {
        let inner = self.inner.clone();
        // JoinError only if the task panics; closing is best-effort.
        let _ = rt().spawn(async move { inner.close().await }).await;
    }
}
